File: internal/inspect/detectors/rmarkdown.go
package detectors

// Copyright (C) 2023 by Posit Software, PBC.

import (
	"fmt"
	"regexp"
	"slices"
	"strings"

	"gopkg.in/yaml.v3"

	"github.com/posit-dev/publisher/internal/config"
	"github.com/posit-dev/publisher/internal/contenttypes"
	"github.com/posit-dev/publisher/internal/executor"
	"github.com/posit-dev/publisher/internal/inspect/dependencies/pydeps"
	"github.com/posit-dev/publisher/internal/logging"
	"github.com/posit-dev/publisher/internal/util"
)

type RMarkdownDetector struct {
	inferenceHelper
	resourceFinderFactory multiResourceFinderFactory
	executor              executor.Executor
	log                   logging.Logger
}

func NewRMarkdownDetector(log logging.Logger) *RMarkdownDetector {
	rfFactory := func(log logging.Logger, base util.AbsolutePath, filesFromConfig []string) (ResourceFinder, error) {
		return NewMultiResourceFinder(log, base, filesFromConfig)
	}

	return &RMarkdownDetector{
		inferenceHelper:       defaultInferenceHelper{},
		resourceFinderFactory: rfFactory,
		executor:              executor.NewExecutor(),
		log:                   log,
	}
}

// Rmd metadata is a YAML block delimited by lines containing only ---
// (this pattern allows it to be followed by optional whitespace)
var rmdMetaRE = regexp.MustCompile(`(?s)^---\s*\n(.*\n)---\s*\n`)

type RMarkdownMetadata struct {
	// Only	the	fields we need are defined here
	Title   string         `yaml:"title"`
	Runtime string         `yaml:"runtime"`
	Params  map[string]any `yaml:"params"`
	Server  any            `yaml:"server"` // string or a map
}

func (d *RMarkdownDetector) getRmdMetadata(rmdContent string) (*RMarkdownMetadata, error) {
	m := rmdMetaRE.FindStringSubmatch(rmdContent)
	if len(m) < 2 {
		return nil, nil
	}
	var metadata RMarkdownMetadata
	decoder := yaml.NewDecoder(strings.NewReader(m[1]))
	err := decoder.Decode(&metadata)
	if err != nil {
		return nil, err
	}
	return &metadata, nil
}

func (d *RMarkdownDetector) getRmdFileMetadata(path util.AbsolutePath) (*RMarkdownMetadata, error) {
	content, err := path.ReadFile()
	if err != nil {
		return nil, err
	}
	return d.getRmdMetadata(string(content))
}

func isShinyRmd(metadata *RMarkdownMetadata) bool {
	if metadata == nil {
		return false
	}
	if strings.HasPrefix(metadata.Runtime, "shiny") {
		return true
	}
	serverString, ok := metadata.Server.(string)
	if ok && serverString == "shiny" {
		return true
	}
	serverMap, ok := metadata.Server.(map[string]any)
	if ok && serverMap["type"] == "shiny" {
		return true
	}
	return false
}

func (d *RMarkdownDetector) isSite(base util.AbsolutePath) (bool, string) {
	for _, ymlFile := range util.KnownSiteYmlConfigFiles {
		exists, err := base.Join(ymlFile).Exists()
		if err == nil && exists {
			d.log.Debug("Site configuration file is present, project being considered as static site", "file", ymlFile)
			return true, ymlFile
		}
	}
	return false, ""
}

func (d *RMarkdownDetector) lookForSiteMetadata(base util.AbsolutePath) (*RMarkdownMetadata, string) {
	// Attempt to get site metadata looking up in common files
	possibleIndexFiles := []string{"index.Rmd", "index.rmd", "app.Rmd", "app.rmd"}
	for _, file := range possibleIndexFiles {
		fileAbsPath := base.Join(file)
		exists, err := fileAbsPath.Exists()
		if err != nil || !exists {
			continue
		}
		metadata, err := d.getRmdFileMetadata(fileAbsPath)
		if err != nil {
			continue
		}
		if metadata != nil {
			return metadata, file
		}
	}
	return nil, ""
}

func (d *RMarkdownDetector) findAndIncludeAssets(base util.AbsolutePath, cfg *config.Config) {
	findAndIncludeAssets(d.log, d.resourceFinderFactory, base, cfg, "RMarkdown")
}

func (d *RMarkdownDetector) configFromFileInspect(base util.AbsolutePath, entrypointPath util.AbsolutePath) (*config.Config, error) {
	relEntrypoint, err := entrypointPath.Rel(base)
	if err != nil {
		return nil, err
	}

	metadata, err := d.getRmdFileMetadata(entrypointPath)
	if err != nil {
		d.log.Warn("Failed to read RMarkdown metadata", "path", entrypointPath, "error", err)
		return nil, err
	}

	cfg := config.New()
	cfg.Entrypoint = relEntrypoint.String()

	if isShinyRmd(metadata) {
		cfg.Type = contenttypes.ContentTypeRMarkdownShiny
	} else {
		cfg.Type = contenttypes.ContentTypeRMarkdown
	}

	if isSite, siteConfigFile := d.isSite(base); isSite {
		var siteMetadata *RMarkdownMetadata
		var indexFile string
		cfg.Files = []string{fmt.Sprint("/", siteConfigFile)}

		if metadata == nil {
			siteMetadata, indexFile = d.lookForSiteMetadata(base)
			metadata = siteMetadata
		}

		if indexFile != "" {
			cfg.Files = append(cfg.Files, fmt.Sprint("/", indexFile))
		}
	}

	// Add the entrypoint to cfg.Files so the resource finder can scan it for assets.
	entrypointFile := fmt.Sprint("/", relEntrypoint.String())
	if !slices.Contains(cfg.Files, entrypointFile) {
		cfg.Files = append(cfg.Files, entrypointFile)
	}

	if metadata != nil {
		title := metadata.Title
		if title != "" {
			cfg.Title = title
		}

		if metadata.Params != nil {
			hasParams := true
			cfg.HasParameters = &hasParams
		}
	}
	needsR, needsPython, err := pydeps.DetectMarkdownLanguages(base)
	if err != nil {
		return nil, err
	}
	if needsR {
		// Indicate that R inspection is needed.
		d.log.Info("RMarkdown: detected R code; configuration will include R")
		cfg.R = &config.R{}
	}
	if needsPython {
		// Indicate that Python inspection is needed.
		d.log.Info("RMarkdown: detected Python code; configuration will include Python")
		cfg.Python = &config.Python{}
	}
	d.findAndIncludeAssets(base, cfg)
	return cfg, nil
}

func (d *RMarkdownDetector) InferType(base util.AbsolutePath, entrypoint util.RelativePath) ([]*config.Config, error) {
	entrypointIsSiteConfig := slices.Contains(util.KnownSiteYmlConfigFiles, strings.ToLower(entrypoint.String()))

	// When the choosen entrypoint is a site configuration yml
	// generate a single configuration as a site project.
	if entrypointIsSiteConfig {
		d.log.Debug("A site configuration file was picked as entrypoint", "entrypoint", entrypoint.String())

		cfg, err := d.configFromFileInspect(base, base.Join(entrypoint.String()))
		if err != nil {
			return nil, err
		}

		if cfg != nil {
			return []*config.Config{cfg}, nil
		}
	}

	if entrypoint.String() != "" {
		// Optimization: skip inspection if there's a specified entrypoint
		// and it's not one of ours.
		if entrypoint.Ext() != ".Rmd" {
			return nil, nil
		}
	}

	var configs []*config.Config
	entrypointPaths, err := base.Glob("*.Rmd")
	if err != nil {
		return nil, err
	}
	for _, entrypointPath := range entrypointPaths {
		relEntrypoint, err := entrypointPath.Rel(base)
		if err != nil {
			return nil, err
		}
		if entrypoint.String() != "" && relEntrypoint != entrypoint {
			// Only inspect the specified file
			continue
		}
		cfg, err := d.configFromFileInspect(base, entrypointPath)
		if err != nil {
			return nil, err
		}
		configs = append(configs, cfg)
	}
	return configs, nil
}

File: extensions/vscode/src/inspect/detectors/rmarkdown.ts
// Copyright (C) 2026 by Posit Software, PBC.

import * as path from "path";
import * as fs from "fs/promises";
import * as yaml from "js-yaml";
import { ContentType } from "src/api/types/configurations";
import { logger } from "src/logging";
import { ContentTypeDetector, PartialConfig } from "../types";
import { globDir } from "../helpers/globDir";
import { detectMarkdownLanguagesInDirectory } from "../helpers/markdownLanguages";

interface RMarkdownMetadata {
  title?: string;
  runtime?: string;
  params?: Record<string, unknown>;
  server?: unknown; // string or { type: string }
}

// Rmd metadata is a YAML block delimited by lines containing only ---
// The .*? is non-greedy to stop at the first closing --- rather than
// matching until a possible later --- (e.g., a horizontal rule in the
// document body).
const rmdMetaRE = /^---\s*\n(.*?\n)---\s*\n/s;

const knownSiteConfigFiles = [
  "_site.yml",
  "_site.yaml",
  "_bookdown.yml",
  "_bookdown.yaml",
];

function getRmdMetadata(content: string): RMarkdownMetadata | undefined {
  const m = rmdMetaRE.exec(content);
  if (!m || !m[1]) {
    return undefined;
  }
  try {
    const metadata = yaml.load(m[1]) as RMarkdownMetadata;
    if (metadata && typeof metadata === "object") {
      return metadata;
    }
  } catch (err: unknown) {
    logger.warn(`[rmarkdown] failed to parse YAML metadata: ${err}`);
  }
  return undefined;
}

function isShinyRmd(metadata: RMarkdownMetadata | undefined): boolean {
  if (!metadata) {
    return false;
  }
  if (metadata.runtime && metadata.runtime.startsWith("shiny")) {
    return true;
  }
  if (typeof metadata.server === "string" && metadata.server === "shiny") {
    return true;
  }
  if (
    typeof metadata.server === "object" &&
    metadata.server !== null &&
    "type" in metadata.server &&
    metadata.server.type === "shiny"
  ) {
    return true;
  }
  return false;
}

export class RMarkdownDetector implements ContentTypeDetector {
  async inferType(
    baseDir: string,
    entrypoint?: string,
  ): Promise<PartialConfig[]> {
    // When the chosen entrypoint is a site configuration yml,
    // generate a single configuration as a site project.
    if (entrypoint && knownSiteConfigFiles.includes(entrypoint.toLowerCase())) {
      logger.debug(
        `[rmarkdown] site configuration file picked as entrypoint: ${entrypoint}`,
      );
      const cfg = await this.configFromFileInspect(baseDir, entrypoint);
      if (cfg) {
        return [cfg];
      }
    }

    if (entrypoint) {
      // Optimization: skip inspection if there's a specified entrypoint
      // and it's not one of ours.
      if (path.extname(entrypoint).toLowerCase() !== ".rmd") {
        return [];
      }
    }

    const configs: PartialConfig[] = [];
    const rmdFiles = await globDir(baseDir, "*.Rmd");

    for (const rmdPath of rmdFiles) {
      const relEntrypoint = path.basename(rmdPath);
      if (entrypoint && relEntrypoint !== entrypoint) {
        continue;
      }
      const cfg = await this.configFromFileInspect(baseDir, relEntrypoint);
      if (cfg) {
        configs.push(cfg);
      }
    }
    return configs;
  }

  private async configFromFileInspect(
    baseDir: string,
    relEntrypoint: string,
  ): Promise<PartialConfig | undefined> {
    const entrypointPath = path.join(baseDir, relEntrypoint);
    let content: string;
    try {
      content = await fs.readFile(entrypointPath, "utf-8");
    } catch (err: unknown) {
      // If the entrypoint is a site config file, we still need to proceed
      // but without file content for metadata extraction.
      logger.warn(
        `[rmarkdown] could not read entrypoint ${entrypointPath}: ${err}`,
      );
      content = "";
    }

    let metadata = getRmdMetadata(content);

    const files: string[] = [];

    // Check if this is a site project
    const siteInfo = await this.findSiteConfig(baseDir);
    if (siteInfo) {
      logger.debug(
        `[rmarkdown] found site configuration file: ${siteInfo.configFile}`,
      );
      files.push(`/${siteInfo.configFile}`);

      if (!metadata) {
        // Look for site metadata in index.Rmd or app.Rmd
        const siteMeta = await this.lookForSiteMetadata(baseDir);
        if (siteMeta) {
          logger.debug(
            `[rmarkdown] found site metadata in fallback file: ${siteMeta.indexFile}`,
          );
          metadata = siteMeta.metadata;
          content = siteMeta.indexContent;
          if (siteMeta.indexFile) {
            files.push(`/${siteMeta.indexFile}`);
          }
        }
      }
    }

    // Add the entrypoint to files
    const entrypointFile = `/${relEntrypoint}`;
    if (!files.includes(entrypointFile)) {
      files.push(entrypointFile);
    }

    const type = isShinyRmd(metadata) ? ContentType.RMD_SHINY : ContentType.RMD;

    const cfg: PartialConfig = {
      type,
      entrypoint: relEntrypoint,
      files,
    };

    if (metadata?.title) {
      cfg.title = metadata.title;
    }

    if (metadata?.params) {
      cfg.hasParameters = true;
    }

    // Detect R and Python code blocks across all .Rmd files in the directory
    const { needsR, needsPython } =
      await detectMarkdownLanguagesInDirectory(baseDir);
    if (needsR) {
      logger.info(`[rmarkdown] detected R code in ${relEntrypoint}`);
      cfg.r = {};
    }
    if (needsPython) {
      logger.info(`[rmarkdown] detected Python code in ${relEntrypoint}`);
      cfg.python = {};
    }

    return cfg;
  }

  private async findSiteConfig(
    baseDir: string,
  ): Promise<{ configFile: string } | undefined> {
    for (const configFile of knownSiteConfigFiles) {
      const configPath = path.join(baseDir, configFile);
      try {
        await fs.access(configPath);
        return { configFile };
      } catch {
        continue;
      }
    }
    return undefined;
  }

  private async lookForSiteMetadata(baseDir: string): Promise<
    | {
        metadata: RMarkdownMetadata;
        indexFile: string;
        indexContent: string;
      }
    | undefined
  > {
    const possibleIndexFiles = ["index.Rmd", "index.rmd", "app.Rmd", "app.rmd"];
    for (const file of possibleIndexFiles) {
      const filePath = path.join(baseDir, file);
      try {
        const content = await fs.readFile(filePath, "utf-8");
        const metadata = getRmdMetadata(content);
        if (metadata) {
          return { metadata, indexFile: file, indexContent: content };
        }
      } catch {
        continue;
      }
    }
    return undefined;
  }
}
